### ISSUE TITLE
Security: add a WAF or similar

### ISSUE THREAD

**Issue by FPMilan** [davestephens/ansible-nas#155]
On my server according to traefik, I get thousands of requests from strangers/bots.
To block all unwanted traffic it would be nice to have some sort of WAF (or Web application firewall) to allow only the wanted traffic. 
I did a little research and could not find a WAF that natively integrates with traefik.
But maybe this https://modsecurity.org/ could work...!?

Has anybody else experience or thought about this?

If this approach is from the wrong angle, please tell me. (e.g. instead of a WAF you might use a pfsense firewall in your network or so)

**Comment by koenvervloesem** [comment]
This is why I would never expose my Ansible-NAS services to internet.

A more secure approach (which I use) would be to:

- run an OpenVPN server (which I do on my OPNsense firewall, but you could easily do this on your Ansible-NAS box as long as you configure port forwarding)
- don't expose your domains to the internet but only in your local network (see https://github.com/davestephens/ansible-nas/issues/128 about how I do this)

You have to connect to your VPN then every time you want to access your services at home from outside. If you secure your VPN connection with a password and optionally a client certificate and/or a one-time password (which you can all easily configure in OPNsense), you can't get more secure than that: there will be no unwanted traffic except some brute-force attempts at your OpenVPN server.

**Comment by FPMilan** [comment]
That is an option for sure! 
Yet I think the risk is not too high if I could restrict access by a firewall, especially if combined with another layer of security like discussed in #125 

Also it is unpractical as my family and friends would have to use a VPN (and by that get access to my entire home network)

**Comment by koenvervloesem** [comment]
In your case you could indeed mitigate the risks by a combination of a web application firewall, Traefik authentication and user management (to selectively give access to specific services to specific users).

But I would probably do the firewall (partially) in the router to block malicious traffic as soon as possible. Also have a look at [Fail2ban](https://www.fail2ban.org/wiki/index.php/Main_Page) to block brute-force attacks.

**Comment by davestephens** [comment]
I guess my main question would be "what problem are you trying to solve?"

By exposing any service on the internet, a side effect you have to accept is that it'll constantly get hit by bots for any number of different attack attempts. Forced login, URL vulnerabilities etc etc. 

Modsecurity could be set up in proxy mode in front of Traefik, but I would imagine it'd be pretty painful to configure in a way that you'd want so it'd actually be useful. Would the effort outweigh the gain? Probably not. 

I'd like to include Pritunl or OpeVPN at some point, which as mentioned above is the best way to ensure proper security without exposing everything to the internet. Fail2ban would also help brute force attempts from single IP addresses.

**Comment by FPMilan** [comment]
What I want is essentially protection against SQL injection, XSS and LFI as well as to automatically block traffic from known malicious IP addresses. 
I think the last one could be kind of done with Fail2ban
Any Idea for the other three?

**Comment by davestephens** [comment]
For which application? Every app will have its own vulnerabilities, so
blanket saying "this is what I want" is not really realistic without being
more specific.

**Comment by FPMilan** [comment]
Did I get it wrong? I thought that a WAF on layer 7 would not need to be specifically configured for every application as it blocks the incoming traffic before it reaches the app.
If this is not how a WAF works I completely got it wrong :D

**Comment by badcrc** [comment]
A valid request for certain app can be an attack vector for another one.

The firewall approach is more efficient imho, banning certain countries works very well. 

PS. Not trying to attack anybody but China, Russia and Turkey have a lot of bots, source: I'm a sysadmin at an ISP.

**Comment by FPMilan** [comment]
If also thought about adding a (virtualized) pfsense firewall to my setup and route the traffic through to block stuff und potentially be able to use a vpn when needed.
Slightly off topic but does anybody here has experience with virtualization on home servers and knows some does and dont's?

**Comment by davestephens** [comment]
If it was me, I wouldn't virtualize the thing that all of the traffic in my
house needed to run through. I want my internet to "just work" so I don't
have to do home IT support when my overly-complex infra craps out!

Just my £0.02 though! :-)

If you are interested in virtualization and your CPU supports it, check out
KVM.

**Comment by FPMilan** [comment]
@davestephens not all traffic would need to go through the virtualized firewall, only the internet facing nas traffic.

KVM looks very promising, I will give it a try ( lets see how much is automatizable with ansible )

**Comment by jesse-troy** [comment]
I wanted to toss in my opinon here and explain how I am handling security and remote connections.

One rule I have is that all of my exposed services are read-only. I have restricted all admin accounts to only be accessible locally. The only exception to this rule is Nextcloud, which supports MFA (and Guacamole, see below).

I was using pritunl (a great vpn btw) for a while, but realized that Guacamole is a better option since I can literally use it anywhere. I replaced the version of Guacamole available in this project with this one which supports MFA: https://hub.docker.com/r/oznu/guacamole

That said, I only expose very few of my services online anyway. Everything runs behind my very basic router which blocks traffic from the web on all ports except 80 and 443 which are forwarded to traefik. 

It's not perfect, but it does strike the correct balance (for me) between being secure and also being usable.
